aptos-move/move-examples/token_objects/guild/sources/guild.move

```diff
@@ -0,0 +1,309 @@
+    }
+
+    #[view]
+    /// Returns the guild token address by name
+    public fun member_token_address(guild_token: Object<GuildToken>, member_token_name: String): address acquires GuildToken {
+        let guild_token_addr = object::object_address(&guild_token);
+        let member_collection_name = &borrow_global<GuildToken>(copy guild_token_addr).member_collection;
+        token::create_token_address(&guild_token_addr, member_collection_name, &member_token_name)
+    }
+
+    /// Adds a guild master to the whitelist. This function allows the admin to add a guild master
+    /// to the whitelist.
+    public entry fun add_guild_master(admin: &signer, guild_master: address) acquires Config {
+        assert!(signer::address_of(admin) == @token_objects, ENOT_ADMIN);
+        let config = borrow_global_mut<Config>(@token_objects);
+        vector::push_back(&mut config.whitelist, guild_master);
+    }
+
+    /// Mints an guild token, and creates a new associated member collection.
+    /// This function allows a whitelisted guild master to mint a new guild token.
+    public entry fun mint_guild(
+        guild_master: &signer,
+        description: String,
+        name: String,
+        uri: String,
+        member_collection_name: String,
+        member_collection_description: String,
+        member_collection_uri: String,
+    ) acquires Config, Manager {
+        // Checks if the guild master is whitelisted.
+        let guild_master_addr = signer::address_of(guild_master);
+        assert!(is_whitelisted(guild_master_addr), ENOT_GUILD_MASTER);
+
+        // The collection name is used to locate the collection object and to create a new token object.
+        let collection = string::utf8(GUILD_COLLECTION_NAME);
+        // Creates the guild token, and get the constructor ref of the token. The constructor ref
+        // is used to generate the refs of the token.
+        let constructor_ref = token::create_named_token(
+            &manager_signer(),
+            collection,
+            description,
+            name,
+            option::none(),
+            uri,
+        );
+
+        // Generates the object signer and the refs. The refs are used to manage the token.
+        let object_signer = object::generate_signer(&constructor_ref);
+        let transfer_ref = object::generate_transfer_ref(&constructor_ref);
+        let extend_ref = object::generate_extend_ref(&constructor_ref);
+
+        // Transfers the token to the `soul_bound_to` address
+        let linear_transfer_ref = object::generate_linear_transfer_ref(&transfer_ref);
+        object::transfer_with_ref(linear_transfer_ref, guild_master_addr);
+
+        // Publishes the GuildToken resource with the refs.
+        let guild_token = GuildToken {
+            extend_ref,
+            member_collection: member_collection_name,
+        };
+        move_to(&object_signer, guild_token);
+
+        // Creates a member collection which is associated to the guild token.
```

Comment:
What are the tradeoffs between this model where guilds are in one collection and members are in another. Vs each guild is one collection?

Reply:
In the current model, there is globally one guild collection, and there are many member collections (one member collection per guild).

Alternatively, there could be a different model, where there could be globally one guild collection and one global member collection. @movekevin, is this model what you're referring to?

The former vs. latter comes down to decentralized member collections vs. the centralized member collection. And the tradeoff may be in the parallelizability and fine-grained controllability (e.g., each member collection could be controlled/managed differently).